pallets/elections/src/impls.rs

```diff
@@ -1,146 +1,74 @@
-    validator_totals
-        .into_iter()
-        .map(|(v, t)| {
-            (
-                v,
-                (t.saturating_mul(MAX_REWARD as u128) / sum_totals) as u32,
-            )
-        })
-        .collect()
 }
 
-fn choose_for_session<T: Clone>(
-    validators: Vec<T>,
-    count: usize,
-    session: usize,
-) -> Option<Vec<T>> {
-    if validators.is_empty() || count == 0 {
-        return None;
-    }
-
-    let validators_len = validators.len();
-    let first_index = session.saturating_mul(count) % validators_len;
-    let mut chosen = Vec::new();
-
-    for i in 0..count.min(validators_len) {
-        chosen.push(validators[first_index.saturating_add(i) % validators_len].clone());
+impl<T: Config> primitives::EraManager for Pallet<T> {
+    fn on_new_era(era: EraIndex) {
+        Self::populate_next_era_validators_on_next_era_start(era);
     }
-
-    Some(chosen)
+    fn new_era_start(_: EraIndex) {}
 }
 
-fn rotate<T: Clone + PartialEq>(
-    current_session: SessionIndex,
-    reserved_seats: usize,
-    non_reserved_seats: usize,
-    reserved: Vec<T>,
-    non_reserved: Vec<T>,
-) -> Option<Vec<T>> {
-    // The validators for the committee at the session `n` are chosen as follow:
-    // 1. `reserved_seats` validators are chosen from the reserved set while `non_reserved_seats` from the non_reserved set.
-    // 2. Given a set of validators the chosen ones are from the range:
-    // `n * seats` to `(n + 1) * seats` where seats is equal to reserved_seats(non_reserved_seats) for reserved(non_reserved) validators.
-
-    let reserved_committee = choose_for_session(reserved, reserved_seats, current_session as usize);
-    let non_reserved_committee =
-        choose_for_session(non_reserved, non_reserved_seats, current_session as usize);
-
-    match (reserved_committee, non_reserved_committee) {
-        (Some(rc), Some(nrc)) => Some(rc.into_iter().chain(nrc.into_iter()).collect()),
-        (Some(rc), _) => Some(rc),
-        (_, Some(nrc)) => Some(nrc),
-        _ => None,
+impl<T: Config> primitives::BanHandler for Pallet<T> {
+    type AccountId = T::AccountId;
+    fn can_ban(account_id: &Self::AccountId) -> bool {
+        !NextEraReservedValidators::<T>::get().contains(account_id)
     }
 }
 
-impl<T> Pallet<T>
+impl<T: Config + pallet_staking::Config + pallet_session::Config> primitives::ValidatorProvider
+    for Pallet<T>
 where
-    T: Config,
+    T::ValidatorId: Into<T::AccountId>,
 {
-    fn update_validator_total_rewards(era: EraIndex) {
-        let validator_totals = T::ValidatorRewardsHandler::validator_totals(era);
-        let scaled_totals = compute_validator_scaled_total_rewards(validator_totals).into_iter();
-
-        ValidatorEraTotalReward::<T>::put(ValidatorTotalRewards(scaled_totals.collect()));
+    type AccountId = T::AccountId;
+    fn current_era_validators() -> Option<EraValidators<Self::AccountId>> {
+        if pallet_staking::ActiveEra::<T>::get().map(|ae| ae.index) == Some(0) {
```

Comment:
Do we really need this check here and below?

Reply:
This is how it was in the election pallet code